mono/metadata/class-init.c

```diff
@@ -403,6 +394,71 @@ mono_class_set_failure_and_error (MonoClass *klass, MonoError *error, const char
 	mono_error_set_type_load_class (error, klass, "%s", msg);
 }
 
+MonoClass *
+mono_class_create_from_invalid_typeref (MonoImage *image, guint32 type_token, MonoError *error)
+{
+	MonoTableInfo *tt = &image->tables [MONO_TABLE_TYPEREF];
+	MonoClass *klass;
+	guint32 cols [MONO_TYPEDEF_SIZE];
+	guint tidx = mono_metadata_token_index (type_token);
+	const char *name, *nspace;
+
+	if (mono_metadata_token_table (type_token) != MONO_TABLE_TYPEREF || tidx > tt->rows) {
+		mono_error_set_bad_image (error, image, "Invalid typeref token %x", type_token);
+		return NULL;
+	}
+
+	mono_loader_lock ();
+
+	if ((klass = (MonoClass *)mono_internal_hash_table_lookup (&image->class_cache, GUINT_TO_POINTER (type_token)))) {
+		mono_loader_unlock ();
+		error_init (error);
+		return klass;
+	}
+
+	mono_metadata_decode_row (tt, tidx - 1, cols, MONO_TYPEREF_SIZE);
+
```

Comment:
Maybe check that `cols [MONO_TYPEREF_SCOPE] & MONO_RESOLUTION_SCOPE_MASK == MONO_RESOLUTION_SCOPE_ASSEMBLYREF` - I'd rather be a bit conservative and only do this for missing references to another assembly, not for nested types or multi-module stuff.